src/main/scala/is/hail/expr/ir/SpecializedArrayBuilders.scala

```diff
@@ -0,0 +1,236 @@
+
+  val ti = typeToTypeInfo(elt)
+
+  val ref: LocalRef[Any] = coerce[Any](ti match {
+    case BooleanInfo => mb.newLocal[BooleanArrayBuilder]("zab")
+    case IntInfo => mb.newLocal[IntArrayBuilder]("iab")
+    case LongInfo => mb.newLocal[LongArrayBuilder]("jab")
+    case FloatInfo => mb.newLocal[FloatArrayBuilder]("fab")
+    case DoubleInfo => mb.newLocal[DoubleArrayBuilder]("dab")
+    case ti => throw new RuntimeException(s"unsupported type found: $elt whose type info is $ti")
+  })
+
+  def create(len: Code[Int]): Code[Unit] = {
+    ref := {
+      ti match {
+        case BooleanInfo => Code.newInstance[BooleanArrayBuilder, Int](len)
+        case IntInfo => Code.newInstance[IntArrayBuilder, Int](len)
+        case LongInfo => Code.newInstance[LongArrayBuilder, Int](len)
+        case FloatInfo => Code.newInstance[FloatArrayBuilder, Int](len)
+        case DoubleInfo => Code.newInstance[DoubleArrayBuilder, Int](len)
+        case ti => throw new RuntimeException(s"unsupported type found: $elt whose type info is $ti")
+      }
+    }
+  }
+
+  def add(i: Code[_]): Code[Unit] = ti match {
+    case BooleanInfo => coerce[BooleanArrayBuilder](ref).invoke[Boolean, Unit]("add", coerce[Boolean](i))
+    case IntInfo => coerce[IntArrayBuilder](ref).invoke[Int, Unit]("add", coerce[Int](i))
+    case LongInfo => coerce[LongArrayBuilder](ref).invoke[Long, Unit]("add", coerce[Long](i))
+    case FloatInfo => coerce[FloatArrayBuilder](ref).invoke[Float, Unit]("add", coerce[Float](i))
+    case DoubleInfo => coerce[DoubleArrayBuilder](ref).invoke[Double, Unit]("add", coerce[Double](i))
+  }
+
+  def apply(i: Code[Int]): Code[_] = ti match {
+    case BooleanInfo => coerce[BooleanArrayBuilder](ref).invoke[Int, Boolean]("apply", i)
+    case IntInfo => coerce[IntArrayBuilder](ref).invoke[Int, Int]("apply", i)
+    case LongInfo => coerce[LongArrayBuilder](ref).invoke[Int, Long]("apply", i)
+    case FloatInfo => coerce[FloatArrayBuilder](ref).invoke[Int, Float]("apply", i)
+    case DoubleInfo => coerce[DoubleArrayBuilder](ref).invoke[Int, Double]("apply", i)
+  }
+
+  def size: Code[Int] = ti match {
+    case BooleanInfo => coerce[BooleanArrayBuilder](ref).invoke[Int]("size")
+    case IntInfo => coerce[IntArrayBuilder](ref).invoke[Int]("size")
+    case LongInfo => coerce[LongArrayBuilder](ref).invoke[Int]("size")
+    case FloatInfo => coerce[FloatArrayBuilder](ref).invoke[Int]("size")
+    case DoubleInfo => coerce[DoubleArrayBuilder](ref).invoke[Int]("size")
+  }
+
+  def clear: Code[Unit] = ti match {
+    case BooleanInfo => coerce[BooleanArrayBuilder](ref).invoke[Unit]("clear")
+    case IntInfo => coerce[IntArrayBuilder](ref).invoke[Unit]("clear")
+    case LongInfo => coerce[LongArrayBuilder](ref).invoke[Unit]("clear")
+    case FloatInfo => coerce[FloatArrayBuilder](ref).invoke[Unit]("clear")
+    case DoubleInfo => coerce[DoubleArrayBuilder](ref).invoke[Unit]("clear")
+  }
+
+}
+
+class ByteArrayBuilder(initialCapacity: Int) {
```

Comment:
can we not use the is.hail.utils.ArrayBuilder because @specialized isn't compatible with our runtime code gen?

Reply:
yeah.

Reply:
(we could write some asm4s stuff to make it compatible, but it involves a lot of manually generating classtags and mucking around with how scala stores specialized classes and other stuff.)

Reply:
suspected so. This seems fine to me.